internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,52 +115,328 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
 
-	requeue := false
 	validatedUpstreams := make([]upstreamprovider.UpstreamGithubIdentityProviderI, 0, len(actualUpstreams))
 	for _, upstream := range actualUpstreams {
-		valid, err := c.validateUpstream(ctx, upstream)
-		if valid == nil {
-			requeue = true
-			errs = append(errs, err)
-		} else {
-			validatedUpstreams = append(validatedUpstreams, upstreamprovider.UpstreamGithubIdentityProviderI(valid))
+		validatedUpstream, applicationErr := c.validateUpstreamAndUpdateConditions(ctx, upstream)
+		if applicationErr != nil {
+			applicationErrors = append(applicationErrors, applicationErr)
+		} else if validatedUpstream != nil {
+			c.log.Info("GitHub Upstream Watcher Sync() found a validated upstream!", "name", validatedUpstream.Name)
+			validatedUpstreams = append(validatedUpstreams, validatedUpstream)
 		}
+		// Else:
+		// If both validatedUpstream and applicationErr are nil, this must be because the upstream had configuration errors.
+		// This controller should take no action until the user has reconfigured the upstream.
 	}
 	c.cache.SetGitHubIdentityProviders(validatedUpstreams)
-	if requeue {
-		return controllerlib.ErrSyntheticRequeue
+
+	// If we have recoverable application errors, let's do a requeue and capture all the applicationErrors too
+	if len(applicationErrors) > 0 {
+		applicationErrors = append([]error{controllerlib.ErrSyntheticRequeue}, applicationErrors...)
+	}
+
+	return errorsutil.NewAggregate(applicationErrors)
+}
+
+func (c *gitHubWatcherController) validateClientSecret(secretName string) (*metav1.Condition, string, string, error) {
+	secret, unableToRetrieveSecretErr := c.secretInformer.Lister().Secrets(c.namespace).Get(secretName)
+
+	// This is not a error that can be recovered with a resync, it requires user interaction.
+	if k8serrors.IsNotFound(unableToRetrieveSecretErr) {
+		unableToRetrieveSecretErr = nil
+	}
+
+	if unableToRetrieveSecretErr == nil &&
+		secret != nil &&
+		secret.Type == gitHubClientSecretType &&
+		len(secret.Data[clientIDDataKey]) > 0 &&
+		len(secret.Data[clientSecretDataKey]) > 0 {
+		return &metav1.Condition{
+			Type:    ClientCredentialsObtained,
+			Status:  metav1.ConditionTrue,
+			Reason:  upstreamwatchers.ReasonSuccess,
+			Message: "client credentials have been obtained, although they may not yet have been validated",
+		}, string(secret.Data[clientIDDataKey]), string(secret.Data[clientSecretDataKey]), nil
+	}
+
+	return &metav1.Condition{
+		Type:   ClientCredentialsObtained,
+		Status: metav1.ConditionFalse,
+		Reason: upstreamwatchers.ReasonNotFound,
+		Message: fmt.Sprintf("secret with name %q must be found in namespace %q with type %q and keys %q and %q",
+			secretName,
+			c.namespace,
+			gitHubClientSecretType,
+			clientIDDataKey,
+			clientSecretDataKey),
+	}, "", "", unableToRetrieveSecretErr
+}
+
+func validateOrganizationsPolicy(upstream *v1alpha1.GitHubIdentityProvider) (*metav1.Condition, v1alpha1.GitHubAllowedAuthOrganizationsPolicy) {
```

Comment:
Should we pass in just the policy, rather than the entire upstream?

Reply:
good idea